pkg/backup/manager.go

```diff
@@ -0,0 +1,313 @@
+func (m *Manager) Search() error {
+	return m.all(
+		[]func() error{
+			m.searchVolumes,
+			m.searchInstances,
+		},
+	)
+}
+
+func (m *Manager) searchVolumes() error {
+	params := &ec2.DescribeVolumesInput{
+		Filters: []*ec2.Filter{
+			&ec2.Filter{
+				Name:   aws.String(fmt.Sprintf("tag:%s", m.BackupTagKey)),
+				Values: []*string{aws.String(m.BackupTagValue)},
+			},
+		},
+		MaxResults: aws.Int64(500),
+	}
+
+	m.logf("Searching for volumes with tag %s:%s\n", m.BackupTagKey, m.BackupTagValue)
+
+	return m.client.DescribeVolumesPages(params,
+		func(page *ec2.DescribeVolumesOutput, lastPage bool) bool {
+			for _, v := range page.Volumes {
+				m.volumes = append(m.volumes, v)
+			}
+			return !lastPage
+		})
+}
+
+func (m *Manager) searchInstances() error {
+	params := &ec2.DescribeInstancesInput{
+		Filters: []*ec2.Filter{
+			&ec2.Filter{
+				Name:   aws.String(fmt.Sprintf("tag:%s", m.ImageTagKey)),
+				Values: []*string{aws.String(m.ImageTagValue)},
+			},
+		},
+		MaxResults: aws.Int64(500),
+	}
+
+	m.logf("Searching for instances with tag %s:%s\n", m.ImageTagKey, m.ImageTagValue)
+
+	return m.client.DescribeInstancesPages(params,
+		func(page *ec2.DescribeInstancesOutput, lastPage bool) bool {
+			for _, r := range page.Reservations {
+				for _, i := range r.Instances {
+					tags := utils.TagSliceToMap(i.Tags)
+					m.logf(
+						"Found instance %s(%s) with matching tag\n",
+						aws.StringValue(i.InstanceId),
+						tags.GetDefault("Name", "<no value>"),
+					)
+					m.instances = append(m.instances, i)
+				}
+			}
+			return !lastPage
+		})
+}
+
+// Backup performs the necessary backups
+func (m *Manager) Backup() error {
+	return m.all(
+		[]func() error{
+			m.backupVolumes,
+			m.backupInstances,
+		},
+	)
+}
+
+func (m *Manager) backupVolumes() error {
+	var wg sync.WaitGroup
+	errorChan := make(chan error, 1)
+
+	for _, volume := range m.volumes {
+		wg.Add(1)
+		go func(v *ec2.Volume) {
+			defer wg.Done()
+			snap, err := m.client.CreateSnapshot(&ec2.CreateSnapshotInput{
+				VolumeId: v.VolumeId,
+			})
+			if err != nil {
+				m.logf("Error creating snapshot for volume '%s'\n", aws.StringValue(v.VolumeId))
+				errorChan <- err
+			} else {
+				m.logf("Created snapshot '%s' for volume '%s'\n",
+					aws.StringValue(snap.SnapshotId),
+					aws.StringValue(v.VolumeId),
+				)
+			}
+		}(volume)
+	}
+
+	wg.Wait()
+
+	select {
+	case err := <-errorChan:
+		return err
+	default:
+	}
+
+	return nil
+}
+
+func (m *Manager) backupInstances() error {
+	var wg sync.WaitGroup
+	errorChan := make(chan error, 1)
+
+	for _, instance := range m.instances {
+		wg.Add(1)
+		go func(i *ec2.Instance) {
+			defer wg.Done()
+			tags := utils.TagSliceToMap(i.Tags)
+			imageName, err := m.formatImageName(i)
+			if err != nil {
+				errorChan <- err
+				return
+			}
+
+			image, err := m.client.CreateImage(&ec2.CreateImageInput{
+				InstanceId: i.InstanceId,
+				Name:       aws.String(imageName),
```

Comment:
Should we explicitly set `NoReboot: aws.Bool(true)` here (our use case), or maybe make it configurable (for other cases)? https://docs.aws.amazon.com/sdk-for-go/api/service/ec2/#EC2.CreateImage

Reply:
I'll add a tag that can be placed on the instance to allow overriding on a case by case basis. If left unspecified on the instance, I'd side with not rebooting.